protocol/contracts/oracle/feed/HybridOracleBase.sol

```diff
@@ -0,0 +1,173 @@
+    AggregatorV3Interface internal _goldFeed;
+
+    bool internal _initialized;
+    uint256 internal _index;
+    uint256 internal _cumulative;
+    uint32 internal _timestamp;
+    uint256 internal _reserve;
+    uint256 internal _reserveMinimum;
+    address internal _esg;
+    address internal _hybridOraclePool;
+
+    bool lastRatioValid;
+    Decimal.D256 lastEsgPerGoldViaBackingAsset;
+
+    constructor (
+        address pair,
+        address esg,
+        address assetBackingUsdOracle,
+        address goldOracle,
+        bytes32 file,
+        uint256 reserveMinimum
+    ) public {
+        _backingAssetPriceFeed = AggregatorV3Interface(assetBackingUsdOracle);
+        _goldFeed = AggregatorV3Interface(goldOracle);
+        _file = file;
+        _pair = IUniswapV2Pair(pair);
+        _reserveMinimum = reserveMinimum;
+        _esg = esg;
+        lastRatioValid = false;
+        lastEsgPerGoldViaBackingAsset = Decimal.one();
+        _hybridOraclePool = Constants.getHybridOraclePoolAddress();
+        setup();
+    }
+
+    function setHybridOraclePoolAddress(address parentAddress) onlyDaoOrTreasury external {
+        Require.that(parentAddress != address(0), _file, "Hybrid Oracle cannot be null");
+        _hybridOraclePool = parentAddress;
+    }
+
+    function setup() private {
+        (address token0, address token1) = (_pair.token0(), _pair.token1());
+        _index = _esg == token0 ? 0 : 1;
+        Require.that(_index == 0 || _esg == token1, _file, "ESG not found");
+    }
+
+    function capture() public onlyHybridOraclePool returns (Decimal.D256 memory, bool) {
+        if (_initialized) {
+            return updateOracle();
+        } else {
+            return initializeOracle();
+        }
+    }
+
+    function initializeOracle() private returns (Decimal.D256 memory, bool) {
+        IUniswapV2Pair pair = _pair;
+        uint256 priceCumulative = _index == 0 ? pair.price0CumulativeLast() : pair.price1CumulativeLast();
+        (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast) = pair.getReserves();
+
+        if (reserve0 != 0 && reserve1 != 0 && blockTimestampLast != 0) {
+            _cumulative = priceCumulative;
+            _timestamp = blockTimestampLast;
+            _initialized = true;
+            _reserve = _index == 0 ? reserve1 : reserve0;
+        }
+
+        lastEsgPerGoldViaBackingAsset = Decimal.one();
+        lastRatioValid = false;
+        return lastCapture();
+    }
+
+    function updateOracle() private returns (Decimal.D256 memory, bool) {
+        Decimal.D256 memory goldPerBackingAsset = updatePrice();
+        uint256 previousReserve = updateReserve();
+        bool valid = !backingAssetReserveTooLow(previousReserve);
+        Decimal.D256 memory goldPriceInUsd = backingAssetUsdPrice().mul(goldPerBackingAsset);
+        lastEsgPerGoldViaBackingAsset = goldPriceInUsd.div(goldPrice());
+        lastRatioValid = valid;
+        return lastCapture();
+    }
```

Comment:
Essentially, get the TWAP of ESG/Backing Asset in the pool, then multiply this by the price of the asset to get the USD value of ESG, then get the ESG per XAU by dividing.